messenger/http_transporter.go

```diff
@@ -261,39 +395,43 @@ func (t *HTTPTransporter) listen() error {
 		t.upid.Port = port
 	}
 
-	t.listener = ln
+	if log.V(3) {
+		t.listener = &loggedListener{ln}
+	} else {
+		t.listener = ln
+	}
 	return nil
 }
 
 // Start starts the http transporter
 func (t *HTTPTransporter) Start() <-chan error {
-	t.lifeLock.Lock()
-	defer t.lifeLock.Unlock()
-
-	select {
-	case <-t.started:
-		// already started
-		return nil
-	case <-t.stopped:
-		defer close(t.started)
-		t.stopped = make(chan struct{})
-		atomic.StoreInt32(&t.stopping, 0)
-	default:
-		panic("not started, not stopped, what am i? how can i start?")
-	}
+	t.stateLock.Lock()
+	defer t.stateLock.Unlock()
+	return t.state.Start()
+}
 
+// start expects to be guarded by stateLock
+func (t *HTTPTransporter) start() <-chan error {
 	ch := make(chan error, 1)
 	if err := t.listen(); err != nil {
 		ch <- err
 	} else {
 		// TODO(yifan): Set read/write deadline.
-		log.Infof("http transport listening on %v", t.listener.Addr())
 		go func() {
-			err := http.Serve(t.listener, t.mux)
+			s := &http.Server{
+				// ReadTimeout: 5 * time.Second,
```

Comment:
probably want ReadTimeout to align with the decoder DefaultReadTimeout: should only impact reading the request-line and headers of the initial request.